text/0033-input-versions.md

```diff
@@ -0,0 +1,166 @@
+```
+{
+    "__cedar::version": "2.4",
+    "contents": [
+        {
+            "uid": { "__entity": { "type": "User", "id": "alice"} },
+            "attrs": {},
+            "parents": [{ "__entity": { "type": "UserGroup", "id": "jane_friends"} }]
+        }, ... 
+    ]
+}
+```
+JSON Cedar schema:
+```
+{
+    "__cedar::version": "2.4",
+    "PhotoApp": {
+        "entityTypes": {
+            "User": {
+                "memberOfTypes": [
+                    "UserGroup"
+                ]
+            }, ...
+        },
+        "actions": ...
+    }
+}
+```
+[Custom Cedar schema](0024-schema-syntax.md):
+```
+@__cedar::version("3.0");
+entity UserGroup;
+entity User in [UserGroup];
+entity List {
+    owner: User
+};
+```
+
+## Motivation
+
+Cedar will occasionally make breaking changes which invalidate prior Cedar input formats. For example [RFC 20](0020-unique-record-keys.md) requires that keys in records be unique, both in policies and in entities, whereas pre-RFC 20 they did not have to be. 
+
+This RFC proposes to define annotations that signal the associated version of a particular input format, so that future versions of Cedar (or tools/services that use Cedar) can handle the situation gracefully. 
+
+For example, by annotating a policy or JSON input file with version 2.4, the parser for Cedar version 3.0 can flag an input with a 2.4-only feature as no longer supported, rather than just failing to parse. Or, a service that uses Cedar can run multiple versions of the Cedar tools, selecting the version that matches the input version they are given.
+
+## Detailed design
+
+This RFC proposes a version annotation on Cedar's main input types: policies, entities (JSON), schema (JSON and custom). 
+
+The design has two main goals:
+
+- **Optional**: No version annotation is required. If none is given, the Cedar parser will assume the most recent version.
+
+- **Uniform**: The way of specifying the version should be similar across all input formats, but also adhere to their particular style.
+
+### Version string
+
+Each input file will be associated with a version string. This is the version the input was written for, e.g., `"2.4.1"` or `"3.0"`. It could be that newer Cedar versions are compatible with older formats, but it is up to newer parsers to know that. For example, a new version of Cedar will add the [`is` operator](0005-is-operator.md) to policies, but an earlier version without that operator may still parse properly. The newer parser can account for this when parsing.
+
+If an input file is annotated with an ill-formed version string, the parser should treat it as a version it does not know about, e.g., as if for a future version of Cedar. It can issue a warning and proceed. If an input file is version-annotated more than once, then the parser should signal an error.
+
+### Policies and Custom Schema
+
+For Cedar policies and custom schema syntax, we propose adding a new top-level declaration:
+```
+__cedar::version(XXX);
```

Comment:
While I like the idea of adding a language version specification to policies, this syntax doesn't look nice or consistent with the syntax within Cedar policies themselves.  To me at least, keeping this consistent with policy syntax is more important than keeping it consistent with JSON. 

How about generalizing the annotation mechanism so that it can apply globally, to a whole policy file, as well as locally, individual policies?  

To my eyes, it would much nicer and more consistent to write just `version("2.4.1");` or `@cedar_version("2.4.1");` at the top of the policy file.  

And we may eventually want to add other global annotations, which would be enabled by a general global annotation mechanism.

Reply:
> I understand that we use `__cedar` in JSON

We don't.  `__cedar` is only proposed to be reserved as a namespace in RFC 24.  It's not currently proposed to be reserved in any other context.  

I think the concern with reserving the `cedar` namespace (in RFC 24 or here) is backwards compatibility.  It's much more likely that existing users are using the `cedar` namespace than the `__cedar` one.

Reply:
Got it; thanks.

In that case, is there any way for us to assess how many users, if any, would be affected if we reserve `cedar`?

I think it would be a pity to give up on the more natural namespace prematurely, especially since we don't know for sure that ` __cedar` is not a breaking change as well.

Reply:
I wholeheartedly agree on aesthethics of the prefix. Strings starting with underscores are.. strange. `cedar` feels a lot more natural. 

~~We should also allow 🌲 emoji as a legitimate alternative to `cedar` as in `🌲::version("1.2.0")`~~ 😇

At this time I am not concerned about backward compatibility. Although we don't have data on this, it seems unlikely that significant number of customers would use either `cedar` for their own purposes. We can provide guidance as part of upgrade instructions.

Reply:
I'll try a poll in Slack.

Reply:
I disagree with `__cedar` being "unnatural" as a reserved namespace item. Macro names starting with `__` are reserved in C and C++, for example. If I saw `__cedar` in a policy, I would immediately think of it as reserved/special, and wouldn't be surprised by the syntax.

I take your point that `java` in `java.lang.Integer` is also "reserved" effectively. However, it hasn't been reserved to this point, and I'm a little concerned about someone already using it. I would be surprised if someone was using it, but I'd be more surprised if someone was using `__cedar`.

Are you also proposing to change RFC 24 so that instead of using `__cedar` for a namespace there, we should use `cedar` instead? Recall:

we propose reserving the namespace `__cedar` and allowing it to be used as a prefix for both primitive and extension types. This way, an user-defined entity type `ipaddr` can be referred to directly, and the extension type `ipaddr` can be referred to as `__cedar::ipaddr`. Likewise, this allows a user to (say) define an entity type `String` that can be be referenced distinctly from `__cedar::String` (though this practice should be discouraged!).

I don't like the idea of using `cedar` in one/some places, and `__cedar` in others. To me it feels better to always use one or the other.

> I understand that we use __cedar in JSON, and that's fine.

Why is the use of `__cedar` fine in the JSON? Why not use `cedar::version` there too?

Reply:
Changed proposed prefix to `@cedar_version` (see RFC for why this works).